### PR TITLE
Add watch label

### PR DESCRIPTION
**Pull request by MarcelMue**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

### CODE REVIEW

File: api/v1alpha4/common_types.go [hunk 1]
```diff
@@ -53,6 +53,9 @@ const (
 	// on the reconciled object.
 	PausedAnnotation = "cluster.x-k8s.io/paused"
 
+	// WatchLabel is the label used to mark the nodes that run on interruptible instances.
```
Comment (stone-z) [inline api/v1alpha4/common_types.go]: This comment should describe the label behavior, looks like a copy-paste mistake from `InterruptibleLabel`

File: main.go [hunk 2]
```diff
@@ -102,6 +103,9 @@ func InitFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&watchNamespace, "namespace", "",
 		"Namespace that the controller watches to reconcile cluster-api objects. If unspecified, the controller watches for cluster-api objects across all namespaces.")
 
+	fs.StringVar(&watchLabelValue, "watch-label", "",
+		"Label value that the controller watches to reconcile cluster-api objects. Label key is always cluster.x-k8s.io/watch. If unspecified, the controller watches for all cluster-api objects.")
```
Comment (stone-z) [inline main.go]: You could use the value here directly

File: util/labels/helpers.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+// HasWatchLabel returns true if the object has a label with the WatchLabel key and a configured value.
+func HasWatchLabel(o metav1.Object, labelValue string) bool {
+	labels := o.GetLabels()
+	if labels == nil {
+		return false
+	}
+	val, ok := labels[clusterv1.WatchLabel]
+	if !ok {
+		return ok
```
Comment (stone-z) [inline util/labels/helpers.go]: Is it a linter thing that this is `ok` and not `false`?

**Review by stone-z — APPROVED** [review]

File: api/v1alpha4/common_types.go (hunk 1 above)
Comment (MarcelMue) [reply api/v1alpha4/common_types.go]: I added a more fitting comment here - feel free to read as my english bad

File: util/labels/helpers.go [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+	}
+	val, ok := labels[clusterv1.WatchLabel]
+	if !ok {
+		return false
+	}
+	if val == labelValue {
+		return true
+	}
+
+	return false
```
Comment (tfussell) [inline util/labels/helpers.go]:
```suggestion
	return val == labelValue
```

File: util/predicates/generic_predicates.go [hunk 5]
```diff
@@ -157,6 +177,12 @@ func ResourceNotPaused(logger logr.Logger) predicate.Funcs {
 	}
 }
 
+// ResourceNotPausedAndHasLabel returns a predicate that returns true only if the
+// ResourceNotPaused and ResourceHasLabel predicates return true.
+func ResourceNotPausedAndHasLabel(logger logr.Logger, labelValue string) predicate.Funcs {
```
Comment (tfussell) [inline util/predicates/generic_predicates.go]: Could this be called `ResourceWatched` or `ResourceNotIgnored` maybe? It's hard to parse the meaning at a glance with the long name.

Comment (MarcelMue) [reply util/predicates/generic_predicates.go]: I followed their precedent here - there were multiple functions like this all following this naming scheme. Currently `All` is completely unused but that's why I chose the name.

Comment (tfussell) [reply util/predicates/generic_predicates.go]: I see. That works then.

File: util/predicates/generic_predicates.go [hunk 6]
```diff
@@ -167,3 +193,19 @@ func processIfNotPaused(logger logr.Logger, obj client.Object) bool {
 	log.V(4).Info("Resource is not paused, will attempt to map resource")
 	return true
 }
+
+func processIfLabelMatch(logger logr.Logger, obj client.Object, labelValue string) bool {
```
Comment (tfussell) [inline util/predicates/generic_predicates.go:197]: Maybe this name should indicate that it's only looking for the watch label or else allow the `HasWatchLabel` function to be passed in as an argument.

File: util/labels/helpers.go [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha4"
+)
+
+// HasWatchLabel returns true if the object has a label with the WatchLabel key and a configured value.
+func HasWatchLabel(o metav1.Object, labelValue string) bool {
+	labels := o.GetLabels()
+	if labels == nil {
+		return false
+	}
+	val, ok := labels[clusterv1.WatchLabel]
```
Comment (tfussell) [inline util/labels/helpers.go]: It's safe to read from a `nil` map so this can be simplified a bit more.
```suggestion
	val, ok := o.GetLabels()[clusterv1.WatchLabel]
```

File: util/labels/helpers.go [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+*/
+
+package labels
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha4"
+)
+
+// HasWatchLabel returns true if the object has a label with the WatchLabel key and a configured value.
```
Comment (tfussell) [inline util/labels/helpers.go]:
```suggestion
// HasWatchLabel returns true if the object has a label with the WatchLabel key matching the given value.
```

File: controllers/cluster_controller.go [hunk 9]
```diff
@@ -64,6 +64,7 @@ const (
 // ClusterReconciler reconciles a Cluster object
 type ClusterReconciler struct {
 	Client client.Client
+	Config
```
Comment (tfussell) [inline controllers/cluster_controller.go]: Since the config only has one field, it might be preferable to pass `WatchLabelValue` into Reconcilers as a field directly (i.e. `Config` -> `WatchLabelValue string`).

Comment (MarcelMue) [reply controllers/cluster_controller.go]: I was going back and forth on this. I agree that it would currently make more sense to just pass the string along.

Comment (MarcelMue) [reply controllers/cluster_controller.go]: I think I will bring this up when I make the upstream PR - let them decide which option they prefer

File: util/predicates/generic_predicates.go (hunk 6 above)
Comment (MarcelMue) [reply util/predicates/generic_predicates.go:197]: Do you have any suggestion for naming? I think adding the function as an argument is going to be unnecessarily complex.
